### DISCUSSION TITLE
Google Benchmark

### DISCUSSION THREAD

**Discussion by manpen** [networkit/networkit#698] [Ideas]
I'm currently in the process of trying to optimise the GraphBuilder and hence are going through some benchmarking code (see PR coming soon I hope). The situation here is, that we have several helper functions such as `measure`, `measureOnce` etc. that are spread around the code base and basically invent the wheel all over again each time.

What's even worse is that each time, the developer has to consider how to properly setup its data structures, average over runtimes, find the right number of iterations for statistically meaningful results and so on. While this is not rocket science, it leads to bugs, inconsistencies and bike-shad discussions.

https://github.com/google/benchmark/ is a library with very similar concepts and syntax to google-test that deals with all these issues. Would you accept a PR to add this as a git submodule in the CMake branch?

**Comment by avdgrinten** [comment]
Would the plan be to use this in future code or to rewrite existing code using Google Benchmark? I would need to take a closer look at it but in principle I am open to accept such PRs.

**Comment by avdgrinten** [comment]
As a side node: A better benchmarking suite for NetworKit which can be run automatically has been on my mind for quite some time. This would require scripts to generate or download large graphs and some scripts to print the benchmarking results. If we had such a suite, it would become much easier to detect performance regressions and/or to optimize the core code (e.g. the graph class).

**Comment by manpen** [comment]
I guess the ultimate goal would be to check (important) and migrate (if sensible) all benchmarks and if possible have some scaling experiments in them; e.g. I was surprised to find that the GraphBuilder's benchmark did not yield any meaningful results (PR follows). However, this is more than a small job. 

What I can do is add some infrastructure similarly to `networkit_add_benchmark` in CMake and to migrate benchmarks which I touch anyway. The remainder should then be investigated separately.
